storage/innobase/lock/lock0lock.cc

```diff
@@ -4692,61 +4756,85 @@ void lock_release_on_prepare(trx_t *trx)
 
   rec_offs_init(offsets_);
 
+  lock_sys.wr_lock(SRW_LOCK_CALL);
+  trx->mutex_lock();
+reiterate:
+  for (lock_t *prev, *lock= UT_LIST_GET_LAST(trx->lock.trx_locks); lock;
+       lock= prev)
   {
-    LockMutexGuard g{SRW_LOCK_CALL};
-    trx->mutex_lock();
-
-    for (lock_t *prev, *lock= UT_LIST_GET_LAST(trx->lock.trx_locks); lock;
-         lock= prev)
+    ut_ad(lock->trx == trx);
+    prev= UT_LIST_GET_PREV(trx_locks, lock);
+    if (!lock->is_table())
     {
-      ut_ad(lock->trx == trx);
-      prev= UT_LIST_GET_PREV(trx_locks, lock);
-      if (!lock->is_table())
+      ut_ad(!lock->index->table->is_temporary());
+      if (!lock->is_rec_granted_exclusive_not_gap())
+        lock_rec_dequeue_from_page(lock, false);
+      else if (UNIV_UNLIKELY(lock->type_mode &
+                             (LOCK_PREDICATE | LOCK_PRDT_PAGE)))
+        /* SPATIAL INDEX locking is broken. */;
+      else
       {
-        ut_ad(!lock->index->table->is_temporary());
-        if (!lock->is_rec_granted_exclusive_not_gap())
-          lock_rec_dequeue_from_page(lock, false);
-        else if (UNIV_UNLIKELY(lock->type_mode &
-                               (LOCK_PREDICATE | LOCK_PRDT_PAGE)))
-          /* SPATIAL INDEX locking is broken. */;
-        else
+        auto cell= lock_sys.rec_hash.cell_get(
+            lock->un_member.rec_lock.page_id.fold());
+        if (lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM))
+          lock_rec_unlock(*cell, lock, PAGE_HEAP_NO_SUPREMUM);
+        else if (unlock_unmodified)
         {
-          auto cell= lock_sys.rec_hash.cell_get(lock->un_member.rec_lock.
-                                                page_id.fold());
-          if (lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM))
-            lock_rec_unlock(*cell, lock, PAGE_HEAP_NO_SUPREMUM);
+          ut_ad(lock->trx->isolation_level > TRX_ISO_READ_COMMITTED ||
+                /* Insert-intention lock is valid for supremum for isolation
+                   level > TRX_ISO_READ_COMMITTED */
+                lock->mode() == LOCK_X ||
+                !lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM));
+          ulint set_nth_bit_calls= trx->lock.set_nth_bit_calls;
+          lock_sys.wr_unlock();
+          trx->mutex_unlock();
+          mtr.start();
+          buf_block_t *block= btr_block_get(
+              *lock->index, lock->un_member.rec_lock.page_id.page_no(),
+              RW_S_LATCH, true, &mtr, nullptr, nullptr
+#if defined(UNIV_DEBUG) || !defined(DBUG_OFF)
+              , BUF_GET_POSSIBLY_FREED
+#endif /* defined(UNIV_DEBUG) || !defined(DBUG_OFF) */
+          );
+          lock_sys.wr_lock(SRW_LOCK_CALL);
+          if (block)
+            lock_rec_unlock_unmodified(block, *cell, lock, offsets, heap);
```

Comment:
Yes, you are absolutely right, this is race condition. I will fix it. But we can't place it in `lock_rec_unlock_unmodified()`. Because in `lock_release_on_prepare()`  it's invoked under `lock_sys.wr_lock()`, but in `lock_release_on_prepare_try()`  it's a combination of `lock_sys.rd_lock()`  + `latch->try_acquire()`. But we can re-get the cell in  `lock_release_on_prepare()`  and `lock_release_on_prepare_try()`  under the correspondent latches after a block is gotten.